File: scaffold/frontend/src/types/voice-events.ts
export type VoiceEventType =
  | "audio.chunk"
  | "control.interrupt"
  | "stt.partial"
  | "stt.final"
  | "assistant.text"
  | "tts.audio"
  | "metrics.turn"
  | "error";

export interface BaseVoiceEvent {
  type: VoiceEventType;
  session_id: string;
  timestamp_ms: number;
}

export interface AudioFormat {
  codec: "pcm_s16le" | "webm_opus";
  sample_rate: number;
  channels: number;
}

export interface AudioChunkEvent extends BaseVoiceEvent {
  type: "audio.chunk";
  seq: number;
  format: AudioFormat;
  audio_base64: string;
}

export interface ControlInterruptEvent extends BaseVoiceEvent {
  type: "control.interrupt";
  turn_id?: string;
  reason: "user_speech_detected" | "user_clicked_stop" | "client_reconnect";
}

export interface SttPartialEvent extends BaseVoiceEvent {
  type: "stt.partial";
  utterance_id: string;
  text: string;
  start_ms: number;
  end_ms: number;
  confidence?: number;
}

export interface SttFinalEvent extends BaseVoiceEvent {
  type: "stt.final";
  utterance_id: string;
  text: string;
  normalized_text: string;
  start_ms: number;
  end_ms: number;
  confidence?: number;
}

export interface AssistantTextEvent extends BaseVoiceEvent {
  type: "assistant.text";
  turn_id: string;
  text: string;
}

export interface TtsAudioEvent extends BaseVoiceEvent {
  type: "tts.audio";
  turn_id: string;
  seq: number;
  codec: "mp3" | "wav" | "pcm_s16le";
  sample_rate: number;
  audio_base64: string;
}

export interface MetricsTurnEvent extends BaseVoiceEvent {
  type: "metrics.turn";
  turn_id: string;
  stt_first_partial_ms?: number;
  stt_final_latency_ms?: number;
  llm_first_token_ms?: number;
  tts_first_audio_ms?: number;
  end_to_end_first_audio_ms?: number;
  interrupt_latency_ms?: number | null;
}

export interface ErrorEvent extends BaseVoiceEvent {
  type: "error";
  code: string;
  message: string;
  retryable: boolean;
  trace_id?: string;
}

export type ClientVoiceEvent = AudioChunkEvent | ControlInterruptEvent;

export type ServerVoiceEvent =
  | SttPartialEvent
  | SttFinalEvent
  | AssistantTextEvent
  | TtsAudioEvent
  | MetricsTurnEvent
  | ErrorEvent;

export type VoiceEvent = ClientVoiceEvent | ServerVoiceEvent;

File: scaffold/backend/app/schemas/voice_events.py
from typing import Literal

from pydantic import BaseModel


class AudioFormat(BaseModel):
    codec: Literal["pcm_s16le", "webm_opus"]
    sample_rate: int
    channels: int


class AudioChunkEvent(BaseModel):
    type: Literal["audio.chunk"] = "audio.chunk"
    session_id: str
    timestamp_ms: int
    seq: int
    format: AudioFormat
    audio_base64: str


class ControlInterruptEvent(BaseModel):
    type: Literal["control.interrupt"] = "control.interrupt"
    session_id: str
    timestamp_ms: int
    turn_id: str | None = None
    reason: Literal["user_speech_detected", "user_clicked_stop", "client_reconnect"]


class SttPartialEvent(BaseModel):
    type: Literal["stt.partial"] = "stt.partial"
    session_id: str
    timestamp_ms: int
    utterance_id: str
    text: str
    start_ms: int
    end_ms: int
    confidence: float | None = None


class SttFinalEvent(BaseModel):
    type: Literal["stt.final"] = "stt.final"
    session_id: str
    timestamp_ms: int
    utterance_id: str
    text: str
    normalized_text: str
    start_ms: int
    end_ms: int
    confidence: float | None = None


class AssistantTextEvent(BaseModel):
    type: Literal["assistant.text"] = "assistant.text"
    session_id: str
    timestamp_ms: int
    turn_id: str
    text: str


class TtsAudioEvent(BaseModel):
    type: Literal["tts.audio"] = "tts.audio"
    session_id: str
    timestamp_ms: int
    turn_id: str
    seq: int
    codec: Literal["mp3", "wav", "pcm_s16le"]
    sample_rate: int
    audio_base64: str


class MetricsTurnEvent(BaseModel):
    type: Literal["metrics.turn"] = "metrics.turn"
    session_id: str
    timestamp_ms: int
    turn_id: str
    stt_first_partial_ms: int | None = None
    stt_final_latency_ms: int | None = None
    llm_first_token_ms: int | None = None
    tts_first_audio_ms: int | None = None
    end_to_end_first_audio_ms: int | None = None
    interrupt_latency_ms: int | None = None


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    session_id: str
    timestamp_ms: int
    code: str
    message: str
    retryable: bool
    trace_id: str | None = None

